src/Umbraco.Core/Events/ScopedNotificationPublisher.cs

```diff
@@ -76,10 +91,7 @@ public void ScopeExit(bool completed)
         {
             if (completed)
             {
-                foreach (INotification notification in _notificationOnScopeCompleted)
-                {
-                    _eventAggregator.Publish(notification);
-                }
+                PublishScopedNotifications(_notificationOnScopeCompleted);
```

Comment:
I've temporarily added the following profiling code around this to check performance improvements:
```c#
using (StaticServiceProvider.Instance?.GetService<IProfilingLogger>()?.TraceDuration(GetType(), "Start PublishScopedNotifications", "Complete PublishScopedNotifications"))
{
    // ...
}
```

When testing on a local setup (SQLite database) using `Umbraco.TheStarterKit` and doing a Deploy workspace restore, both without and with this PR (and similar changes for signature refreshers in Deploy) applied, I get 473 notifications, from 8 different types:
- `ContentSavingNotification` × 90
- `ContentTreeChangeNotification` × 90
- `MediaSavingNotification` × 46
- `MediaSavedNotification` × 46
- `MediaTreeChangeNotification` × 46
- `ContentPublishingNotification` × 30
- `ContentPublishedNotification` × 35

However, before this PR, the profiler timings vary between 550-675ms, but with these changes, they only take between 250-315ms, cutting processing time by more than half! 🚀 This difference will only get bigger with more content that's transferred/restored and an external database (adding additional networking latency/overhead).

Similar improvements should be seen when deploying schema changes, because document type, data type, etc. changes trigger a complete NuCache rebuild and Examine reindexing...

Reply:
This is a great PR @ronaldbarendse! In regards to your note about backwards compatibility of the JSON format, the counts can most likely be converted to a faux get/set property: Don't actually set an inner property and always get the length of the Array.

I do wonder how bad breaking this is though... Cache instructions are just that, wipe the table, rebuild your caches and everything is up to date again.

Reply:
Probably out of scope for this particular PR, but instead of sending 473 notifications, would be great if those can also be batched as a whole, especially since Umbraco can decide to stop handling the cache notifications and just rebuild its whole cache.

Reply:
Thanks Lennard! I've managed to get the instructions JSON a bit smaller again without breaking compatibility and I'm leaving the bigger tasks for later, so they can be combined with other (breaking) changes 👍🏻 

All notifications within a scope are already batched in the `IScopeNotificationPublisher`, which you specify when creating an Umbraco scope. But you probably don't want this implementation to be tightly coupled to cache rebuilds (application caches, NuCache, Examine indexes and potentially other custom caches), hence the grouping of notifications and filtering by handler type. This keeps it decoupled, while still batching notifications (by type) 😄